### ISSUE TITLE
SelectCheckboxMenu: Update accessibility

### ISSUE THREAD

**Issue by Rene-Stoeckel** [primefaces/primefaces#11850]
### Description

When the SelectCheckboxMenu is used with screenreaders the following issues can be observed:
1) The widget is attributed with role="combobox", but neither supports the typing of text nor navigation with arrow keys, as
expected in the official  "Combobox Pattern | APG | WAI | W3C", see https://www.w3.org/WAI/ARIA/apg/patterns/combobox/
2) The element with the fokus is marked with readonly="readonly", although it is intended to be changed.
3) The subelements are marked with aria-hidden=true so that the screenreader does not read them when navigating via tab.
4) The items in the SelectCheckboxMenu are labelled in a way so that their label is read twice by the screenreader and their role is read four times by the screen reader.
5) When the popup list is opened the fokus is not set on the popup list but remains on the input.

For reference: The Primefaces SelectCheckboxMenu is rendered into the following structure:
```
<td role="gridcell" class="ui-panelgrid-cell">
	<label id="main:j_id_28" class="ui-outputlabel ui-widget" for="main:status_focus">
		<span class="ui-outputlabel-label">Status:</span>
	</label>
</td>
<td role="gridcell" class="ui-panelgrid-cell">
	<div id="main:status" class="ui-selectcheckboxmenu-multiple ui-selectcheckboxmenu ui-widget ui-state-default ui-corner-all" role="combobox" aria-haspopup="listbox" aria-expanded="false">
		<div class="ui-helper-hidden-accessible">
			<input id="main:status_focus" name="main:status_focus" type="text" readonly="readonly" role="combobox" aria-hidden="true" aria-controls="main:status_panel" aria-expanded="false">
		</div>
		<div class="ui-helper-hidden">
			<input id="main:status:0" name="main:status" type="checkbox" value="ALL" data-escaped="true" aria-checked="aria-checked" checked="checked" aria-labelledby="main:j_id_28">
			<label for="main:status:0">ALL</label>
			<input id="main:status:1" name="main:status" type="checkbox" value="APPROVED" data-escaped="true" aria-labelledby="main:j_id_28">
			<label for="main:status:1">APPROVED</label>
			<input id="main:status:2" name="main:status" type="checkbox" value="CANCELED" data-escaped="true" aria-labelledby="main:j_id_28">
			<label for="main:status:2">CANCELED</label>
```

### Describe the solution you would like

1) Remove the aria role "combobox" as this widget works very differently and 
does not implement the official  "Combobox Pattern | APG | WAI | W3C", 
see https://www.w3.org/WAI/ARIA/apg/patterns/combobox/
2) Use a button to trigger the popup list with the elements instead a div.
3) Link this button to the elements by using a aria-describedby attribute on the items.
4) Let the fokus be set to the popup list when it is openened.

### Additional context

Primefaces 13.0.1 - 13.0.8.

**Comment by melloware** [comment]
PR is welcome...

**Comment by melloware** [comment]
@Rene-Stoeckel i may be able to help with this one.

**Comment by melloware** [comment]
Hmmm the PrimeVue MultiSelect uses ARIA role `combobox` see: https://primevue.org/multiselect/#accessibility

Can you try PrimeVue and see if it behaves with the screenreader how you expect it to?

**Comment by Rene-Stoeckel** [comment]
On a first guess I suspec this multiselect would not be usable with screenreader as well, because the browser fokus remains on the input with role="combobox" (fokus trap, same issue as on the data table) while navigating through the options on the listbox. But I try to get feedback from a tester with JAWS on this.

**Comment by melloware** [comment]
Great that is helpful. I will take a deeper look.

**Comment by Rene-Stoeckel** [comment]
Here is some feedback with JAWS on the demo pf-11850.zip:

1. The popup opens with space or enter or alt+cursorDown, but not cursor-down or cursor-up as defined in the combobox pattern and expressed by the screenreader.
2. The fokus is set to the popup item list when it is opened.
3. Navigation is possible via tab and shift-tab, but cursor up and down should be used primarily according to the combobox pattern. 
4. The selection/deselection of items only works with shift-space where screenreader correctly confirms the activation or the deactivation of a checkbox. When using space only (or enter) the item gets activated visually, but the screenreader gives no confirmation and reports the item as deactivated when navigating back. I'ts unclear if this is a JAWS issue, but it's not observed on the primevue multiselect. 
5. When navigating from each item to the other via tab the screenreader reads the label of the item and the state, but not the number of the item and the out of total number as expected and shown in the primevue demo. 
6. Once the fokus is on the closer icon, then its possible to step through the options with the cursor keys so that this cursor keys navigation does not scroll the currently focussed item into visibility on the popup panel, in contrast to tab (checkbox to checkbox) navigation. It's also not clear if this is a JAWS issue because it seems to add this kind of cursor key navigation due to the missing cursor-keys handling. 
7. The screenreader correctly reads the selected items of the combobox when the popup is closed, but on the screen only "Item(s) Selected" is displayed and the labels of the items are not there.

To summarize, it's an improvement (2 and 7). It's unclear if the focus trap is a problem, because the primevue multiselect navigation works as expected with the screenreader even though the fokus stays on the same input all time there as well. JAWS probably just deals fine with it because of the correct use of the aria attributes.

**Comment by melloware** [comment]
Great feedback!  Let me see if I can fix a few more of these like the cursor handling and better keyboard support.

as for #7 i always make the screen reader value all the selected values.  the visual display of `"Item(s) Selected""` is because of the property `selectedLabel="Item(s) Selected"` i wasn't sure if you wanted that to override the "selected" value for visually impaired users?

**Comment by melloware** [comment]
@Rene-Stoeckel can you have your team test this version: http://20.161.11.140/showcase/ui/input/checkboxMenu.xhtml

I have made most of the changes you requested above and tested some with NVDA and it seems to be working for the most part.  Let me know how it goes with JAWS.

**Comment by Rene-Stoeckel** [comment]
We took a look on the SelectCheckboxMenu showcase with JAWS and here are our observations. Basic, Multiple and Grouped have significantly improved already and there are only minor findings.

Basic / Multiple / Group
- The options listbox is opened with cursor keys and cusor navigation works. 
- However, when the listbox is opened with space then the first item is always toggled unintentially.
- While navigating with the screenreader through the options with cursor keys (or tab) the screenreader correctly reads the item with name and position, e.g.  "Paris 3 of 9". But the screenreader does not tell if the item under the cursor is activated or not.
- The listbox with the options can be closed only when pressing ESC twice.
- The toggler to select none / all entries can only be toggled with shift space, not space as expected.
- The Screenreader correctly reads the selected items when options listbox is closed.
- Under [Basic]: instead of the labels of the selected items the text "Item(s) Selected" is displayed.

Advanced:

- The options listbox opens with cursor keys, but is not usable with the cursor keys until any checkbox is selected with the mouse first, which probably just sets the fokus.
- When navigating with tab to a checkbox of an item e.g. "algeria" the screenreader reads "english algeria, algeria, english algeria controlfield not activated" (algeria mentioned 3 times). But it does not tell position out of (total number of items).
- when navigating with cursor keys to a checkbox of an item, e.g. "algeria" the screenreader reads "algeria controlfield not activated". A keypress on cursor down navigates to the label right to the flag and screen reader reads "algeria listentry 2 of 249". The checkbox or list item the screenreader just reads is not drawn with the blue fokus / highlight class in this situation. The focus highlight remains on the checkbox that was lastly activated via tab. Another key press on cursor down navigates to the checkbox of entry "American Samoa". This navigation with cursor keys does not scroll the options listbox, but closes the listbox instead. This strange navigation effect is only seen with active screenreader JAWS.
- Space on an item deactivates the item visually, but the screenreader tells the item is still activated. Shift+space seems to work. 
- When an item is deselected, the blue fokus marker of the checkbox is removed.
- When the options listbox is closed, the screenreader reads its value "68138" instead of the text of the selected item "algeria".

**Comment by melloware** [comment]
Good feedback. Let me do some more improvements.

**Comment by melloware** [comment]
Also I have noticed I think that is how the screen reader works

"Paris 3 of 9"
"Paris Uneselected 3 of 9"

I think if it doesn't read unselected you assume it's selected

**Comment by Rene-Stoeckel** [comment]
Normally the screenreader would tell the state of an item (checkbox) when fokus enters no matter if its selected or not. Here it does'nt recognize the checkboxes as unselected as well.

**Comment by melloware** [comment]
@Rene-Stoeckel updated showcase: http://20.161.11.140/showcase/ui/input/checkboxMenu.xhtml

some notes:

**BASIC:**
- **FIXED**: Pressing <kbd>SPACE</kbd> key now does not select the first element of the popup
- **FIXED**: Pressing <kbd>SPACE</kbd> key on Select ALL now works
- **FIXED**: Label now reads "Cities Selected" or whatever your selectedLabel="" is set to...
- **EXPECTED**: Pressing <kbd>ESC</kbd>  key twice is expected screen reader behavior see: https://github.com/nvaccess/nvda/issues/4428#issuecomment-155326950 First ESC gets you out of "input mode" second ESC closes the dialog.  You can use <kbd>SHIFT+ARROW_UP</kbd> to close immediately.
- **EXPECTED**: Reading "Selected vs Not Selected" is the expected screen reader behavior see: https://stackoverflow.com/a/71475345

> role="option" default is "not selected" whether you have the `aria-selected` attribute or not....only get [silence, meaning "selected"] when `aria-selected=true`


### ADVANCED:
- **FIXED**: First item gets focus
- **FIXED**: All navigation keys
- **FIXED**: <kbd>SPACE</kbd>now selects item
- **FIXED**: Focus now remains when unselecting an item.
- **FIXED**: Reading correct "Algeria 1 of 249"

**Comment by Rene-Stoeckel** [comment]
Alright, thank you, I hope I can give feedback tomorrow.

**Comment by Rene-Stoeckel** [comment]
Hello, we have some more feedback from JAWS testing:

Major problems:

1) When navigating with tab to the combobox the input element receives the fokus, which is attributed with aria-hidden=true and whould according to the spec not be reported by the screenreader. To be correct here the fokus should be transferred to the div element with the role=combobox (see screenshot): ![11850-role-combobox](https://github.com/primefaces/primefaces/assets/102955224/c5b1e62e-6f84-444c-bf66-4de5d0bb4a17)

2) The item list in the dialog is attributed with role=group. Therefor it is not reported as item list. When using JAWS the status if the item (selected or no selected) is not reported. Rather the list should be attributed with role=listbox aria-multiselectable=true and the list items should carry the attribute role=option. See Screenshots: 
![11850-a](https://github.com/primefaces/primefaces/assets/102955224/d2465066-b092-462f-98f6-92f02ccb23af)
![11850-b](https://github.com/primefaces/primefaces/assets/102955224/4b5f0a29-cf39-4f94-9546-467245d0a284)
![11850-c](https://github.com/primefaces/primefaces/assets/102955224/12132f37-c2ac-4067-aac1-52b6995a003e)

3) When list items are filtered then their total and position is reported incorrectly (not adjusted to the filter) by the screenreader: See Screenshot: 
![11850-d](https://github.com/primefaces/primefaces/assets/102955224/406534fd-34d7-4abd-9a77-24c2e0e23a16)

Minor issues:
•	The selected values in the closed combobox are not reported with NVDA when the page is read using cursor keys or the element overview (INS+F7) is used. The selected values should be accessible by any navigation method.
•	Within the the item list the curser keys can be used for navigation, as well as tab. It is recommended to use only one correct navigation method according the its role. When using item lists with role=listbox its the cursor key navigation, on a checbox group it should only be tab.
•	When the filter is used then the total number of filter results is not reported by the screen reader. It is recommended to use a live region to report the correct count.
•	The opened listbox dialog has no label.
•	The order of the Fokus traversal is irritating (filter-field, closer and the the item list). The close button should be either on the beginning or the end of the fokus sequence.
•	Because of the wrong attributation of the input field (an input within a div, both with role=combobox, see point 1 above) JAWS is changing from linear reading to the form mode and openes the dialog. Thus a fluent reading of the page is impossible without interacting with the componentn unwillingly.
•	The checkbox „All items selected/unselected“ cannot be operated with JAWS and NVDA
•	The checkbox „All items“ shows the selection status programmatically and also in text form (as „activated“ or „not activated). This is redundant, e.g. „All items unselected, checkbox not activated“ both means that nothing was selected.

**Comment by melloware** [comment]
@Rene-Stoeckel all changes made (I think) now it works much more like PrimeVue.

See: http://20.161.11.140/showcase/ui/input/checkboxMenu.xhtml

- **FIXED:** Removed aria-hidden="true" from keyboard trap
- **FIXED**: Removed role="dialog" from popup
- **FIXED**: role="combobox" now only listed once with proper aria-haspopup="listbox"
- **FIXED**: UL now has role="listbox" and aria-multiselectable="true"
- **FIXED**: TAB key now cycles only among the few elements
- **FIXED**: Removed TAB key from listbox and only allow cursor keys
- **FIXED**: Filtering now updates ARIA "positions" 1 of 4 etc
- **FIXED**: Select All checkbox remove aria-selected

**Comment by melloware** [comment]
@Rene-Stoeckel any update on testing this?

**Comment by Rene-Stoeckel** [comment]
Here is new feedback from our test team:

Most of the problems are solved, just these are remaining:

- the sequence of the fokus in the dialog is confusing ("All items" checkbox, close button, options list). The close button should be on the beginning or the end of the dialog.
- the "all items" checkbox can not be accessed with JAWS and NVDA, see below why.
- the checkbox "all items" reports its status to the screen reader programmatically and by text (it sais "all items unselected control field not activated" which is confusing.
- although the combobox role was removed from the outer div, JAWS automaticall switches from linear reading mode to the formular mode and opens the dialog. Reading the page without interacting with the combocox unwillingly is not possible.
- when the filter is applied it shows still the number of total items instead (as expected) the number of filtered items. this is just a usability thing.

 
New observations:
- because the input field is no longer attributed as aria-hidden it is reported correctly by the screen reader. JAWS however reports that (according to the cobobox spec) text could be entered freely, which is not possible. To be perfect this would need to be implemented, but a proper documentation would be acceptable as a workaround. 
- the role was removed from the popup which was not requested, the criticism was that it has no descriptive label. the role was corret and is needed via ARIA specification, but although its missing there seems to be no negative side effect, so its just a hint.

As all major issues were solved, using the component is acceptable from an accessibility standpoint although there are still these open issues menioned above.

>>>What needs to be changed to make the checkbox "all items" usable by the screen reader?

"Most likely just the JavaScript event handlers. For certain control elements, screen readers automatically convert a key event (such as the space bar and enter key) into a click event (e.g. because the space bar on a checkbox is identical to a mouse click). If only a key event is defined at the checkbox and not a click event, then it won't work with a screen reader. As soon as I deactivated the event conversion in the screen reader, I can also operate the checkbox with the screen reader. The solution would be not to use events at all (because the HTML checkbox can operate it on its own) or to always implement key and click events at the same time when using events."

**Comment by melloware** [comment]
@Rene-Stoeckel i just redeployed: http://20.161.11.140/showcase/ui/input/checkboxMenu.xhtml?jfwid=2fd24

1. I put `role="dialog"` back on the popup
2. I think I fixed the "Select All Items" checkbox based on your findings it seems to work for me in NVDA but let me know if JAWS is fixed.
3. I changed the TAB order of elements to make the Close button last.   But let me know the order you think the TAB should be!
